### PR TITLE
Fixes #219

### LINKED ISSUE #219: ADD and ENV sub-instruction of Docker ONBUILD instruction not considered while preparing build-context for building image from Dockerfile
Example Dockerfile snippet -

--Set build triggers for adding contents of artifacts, scripts and tests directory to image while building image for component
ONBUILD ADD $SOURCE_ARTIFACT_DIR $ARTIFACT_DIR/
ONBUILD ADD $SOURCE_SCRIPT_DIR $SCRIPT_DIR/
ONBUILD ADD $SOURCE_TEST_DIR $TEST_DIR/

--Set build trigger for copying the articats(jar / war / ear) to tomcat webapps directory from artifact directory
ONBUILD RUN cp $ARTIFACT_DIR/* $TOMCAT_HOME/webapps/

**************************************************************************************
Following code-block in docker-java is not considering the fact that the base image used in the FROM instruction of a Dockerfile may have onbuild instructions like the ones above

Class : Dockerfile$ScannedResult
Default constuctor : Dockerfile$ScannedResult()

Code snippet : 
for (DockerfileStatement statement : getStatements()) {
        if (statement instanceof DockerfileStatement.Env) {
          processEnvStatement((DockerfileStatement.Env) statement);
        } else if (statement instanceof DockerfileStatement.Add) {
          processAddStatement((DockerfileStatement.Add) statement);
        }
}

In the code above only ADD and ENV instructions in the current docker file are being considered.

However the base image used in the Dockerfile may also contain ADD / ENV instructions as part of ONBUILD instructions.

For the time being , I have commented out the code in my local workspace and replaced it with the following code logic -

// adding all files and subfolders in context folder (where Dockerfile is located) to the build context
// this is how docker build works at present
addAllFilesToContext(dockerFile.getParentFile());

// add input folder and all subfolders and files to the file list
private void addAllFilesToContext(File dockerContextFolder) {
		
    	File[] files = dockerContextFolder.listFiles();
    	
    	if(files != null && files.length > 0) {
    		
    		for(File file : files) {
    			
    			filesToAdd.add(file);
    			
    			if(file.isDirectory()) {
    				
    				addAllFilesToContext(file);
    				
    			}
    			
    		}
    		
    	}
		
}

### PR DESCRIPTION
**Pull request by orzeh**
I have rewritten the logic that adds files to the build context: now it adds __all files__ in `Dockerfile` parent directory, except those excluded by `.dockerignore`.

I have also added some tests that demonstrates the issue, which use `ONBUILD` in `Dockerfile`.

I also fixed small bug in `.dockerignore` test: exclude pattern was wrong according to the [Dockerfile doc](http://docs.docker.com/engine/reference/builder/#dockerignore-file).

### CODE REVIEW

**Comment by marcuslinke** [comment]
@orzeh Adding all files to the build context could be problematic if there are huge files. I really don't know how docker CLI handles this, but seems the builder inspects metadata of the base image (http://docs.docker.com/engine/reference/builder/#onbuild):

> At the end of the build, a list of all triggers is stored in the image manifest, under the key OnBuild. They can be inspected with the docker inspect command.

> Later the image may be used as a base for a new build, using the FROM instruction. As part of processing the FROM instruction, the downstream builder looks for ONBUILD triggers, and executes them in the same order they were registered. If any of the triggers fail, the FROM instruction is aborted which in turn causes the build to fail. If all triggers succeed, the FROM instruction completes and the build continues as usual.

**Comment by marcuslinke** [comment]
@orzeh After thinking about again probably you're right. So uploading all files and ignore only those that matches patterns in `.dockerignore`. Seems our original implementation that adds only files mentioned in the `Dockerfile` was some kind of premature optimization then.

**Comment by orzeh** [comment]
@marcuslinke  I agree with you. 

If one has to deal with huge files, he can use `.dockerignore` or - to take full control over what is send to docker daemon - provide own InputStream via [BuildImageCmd.withTarInputStream()](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java#L122).

**Comment by KostyaSha** [comment]
Wouldn't better place Dockerfile in separate directory?

**Comment by orzeh** [comment]
@KostyaSha which `Dockerfile` do you mean?

**Comment by KostyaSha** [comment]
Any :) from my experience i always placing Dockerfile in clean directory to not have accidental uploads to build context.

**Comment by KostyaSha** [comment]
In general PR LGTM :+1:  even i not fully understand ONBUILD feature and all possible cases :)

**Comment by orzeh** [comment]
Ok, now I get it.

I use ONBUILD when I have several Spring Boot microservices and I add them to the image as a _exploded JAR_ to leverage Docker caching (see [this blog post](http://bsideup.blogspot.in/2015/04/spring-boots-fat-jars-vs-docker.html)). Then I create base image with number of `ONBUILD ADD ...` instructions and just put `FROM my-base-image` in microservices Dockerfiles.

File: src/test/resources/testDockerignore/.dockerignore [hunk 1]
```diff
@@ -1 +1 @@
-b
\ No newline at end of file
+*/b
```
Comment (marcuslinke) [inline src/test/resources/testDockerignore/.dockerignore]: @orzeh Could you explain this change please?

Comment (orzeh) [inline src/test/resources/testDockerignore/.dockerignore]: According to `.dockerignore` doc:
> the root of the context is considered to be both the working and the root directory

then pattern `b` excludes all files and directories that are named b in root directory of build context.
However, in [this test](https://github.com/docker-java/docker-java/blob/master/src/test/java/com/github/dockerjava/core/command/BuildImageCmdImplTest.java#L191) the file named b is under directory a, so it should not be ignored. After my changes this tests fails.

Pattern `*/b`  works as expected (and described in the [doc](http://docs.docker.com/engine/reference/builder/#dockerignore-file) see `*/temp*` example)

Comment (marcuslinke) [inline src/test/resources/testDockerignore/.dockerignore]: OK, now I got it. Thanks for explanation.

**Comment by marcuslinke** [comment]
@orzeh Thanks for contributing!

**Pull request merged**